<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ резюме из HeadHunter
   

In [2]:
import pandas as pd
import numpy as np
import re
import plotly.express as px
from itertools import chain
from IPython.display import IFrame

# Исследование структуры данных

## Чтение данных
Прочитайте данные с помощью библиотеки Pandas. Совет: перед чтением обратите внимание на разделитель внутри файла.

In [3]:
df_raw = pd.read_csv("dst-3.0_16_1_hh_database.csv", sep=";")

In [4]:
df = df_raw.copy(deep=True)

## Вывод данных
Выведите несколько первых (последних) строк таблицы, чтобы убедиться в том, что ваши данные не повреждены. Ознакомьтесь с признаками и их структурой.

In [5]:
df.tail(5)

Пол, возраст           ЗП  \
44739    Мужчина ,  30 лет , родился 17 января 1989   50000 руб.   
44740      Мужчина ,  27 лет , родился 5 марта 1992   39000 руб.   
44741  Женщина ,  48 лет , родилась 26 декабря 1970   40000 руб.   
44742   Мужчина ,  24 года , родился 6 октября 1994   20000 руб.   
44743    Мужчина ,  38 лет , родился 25 апреля 1980  120000 руб.   

                               Ищет работу на должность:  \
44739  Финансист, аналитик, экономист, бухгалтер, мен...   
44740             Системный администратор, IT-специалист   
44741                         Аналитик данных, Математик   
44742                                   Контент-менеджер   
44743                               Руководитель проекта   

                            Город, переезд, командировки  \
44739  Тверь , готов к переезду (Москва, Химки) , гот...   
44740  Липецк , готов к переезду , готов к командировкам   
44741  Челябинск , готова к переезду , готова к редки...   
44742  Тамбов , не готов к переезду , не готов к кома...   
44743  Москва , не готов к переезду , не готов к кома...   

                                               Занятость  \
44739                                   полная занятость   
44740  проектная работа, частичная занятость, полная ...   
44741                                   полная занятость   
44742              частичная занятость, полная занятость   
44743                                   полная занятость   

                                                  График  \
44739                      полный день, удаленная работа   
44740  удаленная работа, гибкий график, полный день, ...   
44741                      полный день, удаленная работа   
44742                                   удаленная работа   
44743                                        полный день   

                                             Опыт работы  \
44739  Опыт работы 7 лет 7 месяцев  Финансист, аналит...   
44740  Опыт работы 7 лет  Системный администратор, IT...   
44741  Опыт работы 21 год 5 месяцев  Январь 1998 — по...   
44742  Опыт работы 3 года 10 месяцев  Контент-менедже...   
44743  Опыт работы 15 лет 10 месяцев  Руководитель пр...   

                         Последнее/нынешнее место работы  \
44739  ООО "IAS" (независимый участник объединения Ru...   
44740                                       ИП Пестрецов   
44741  ОАО «ЧМК», Исследовательско-Технологический Це...   
44742                                          IQ-Maxima   
44743                                  ПАО ГК ТНС энерго   

                    Последняя/нынешняя должность  \
44739                     Руководитель субгруппы   
44740                            Предприниматель   
44741                 Начальник группы аналитики   
44742                          Менеджер проектов   
44743  Руководитель отдела технической поддержки   

                                       Образование и ВУЗ Обновление резюме  \
44739  Высшее образование 2015  Московский гуманитарн...  22.04.2019 12:32   
44740  Высшее образование (Бакалавр) 2016  Воронежски...  22.04.2019 13:11   
44741  Высшее образование 2000  Южно-Уральский госуда...  09.04.2019 05:07   
44742  Высшее образование 2015  Тамбовский государств...  26.04.2019 14:25   
44743  Высшее образование 1997  Южно-Российский госуд...  05.07.2018 20:15   

                                 Авто  
44739                      Не указано  
44740                      Не указано  
44741                      Не указано  
44742  Имеется собственный автомобиль  
44743                      Не указано

## Основная информация
Выведите основную информацию о числе непустых значений в столбцах и их типах в таблице.

Обратите внимание на информацию о числе непустых значений.

In [6]:
df.shape

(44744, 12)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44744 entries, 0 to 44743
Data columns (total 12 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   Пол, возраст                     44744 non-null  object
 1   ЗП                               44744 non-null  object
 2   Ищет работу на должность:        44744 non-null  object
 3   Город, переезд, командировки     44744 non-null  object
 4   Занятость                        44744 non-null  object
 5   График                           44744 non-null  object
 6   Опыт работы                      44576 non-null  object
 7   Последнее/нынешнее место работы  44743 non-null  object
 8   Последняя/нынешняя должность     44742 non-null  object
 9   Образование и ВУЗ                44744 non-null  object
 10  Обновление резюме                44744 non-null  object
 11  Авто                             44744 non-null  object
dtypes: object(12)
memory usage: 4.1+

## Основная статистическая информацая
Выведите основную статистическую информацию о столбцах.


In [8]:
df.describe()

Пол, возраст          ЗП  \
count                                           44744       44744   
unique                                          16003         690   
top     Мужчина ,  32 года , родился 17 сентября 1986  50000 руб.   
freq                                               18        4064   

       Ищет работу на должность:  \
count                      44744   
unique                     14929   
top      Системный администратор   
freq                        3099   

                             Город, переезд, командировки         Занятость  \
count                                               44744             44744   
unique                                              10063                38   
top     Москва , не готов к переезду , не готов к кома...  полная занятость   
freq                                                 1261             30026   

             График                                        Опыт работы  \
count         44744                                              44576   
unique           47                                              44413   
top     полный день  Опыт работы 10 лет 8 месяцев  Апрель 2018 — по...   
freq          22727                                                  3   

                          Последнее/нынешнее место работы  \
count                                               44743   
unique                                              30214   
top     Индивидуальное предпринимательство / частная п...   
freq                                                  935   

       Последняя/нынешняя должность  \
count                         44742   
unique                        16927   
top         Системный администратор   
freq                           2062   

                                        Образование и ВУЗ Обновление резюме  \
count                                               44744             44744   
unique                                              40148             18838   
top     Высшее образование 2014 Тюменский Государствен...  07.05.2019 09:50   
freq                                                    4                25   

              Авто  
count        44744  
unique           2  
top     Не указано  
freq         32268

###  Задание 2.1
Прочитайте данные с помощью библиотеки Pandas. Перед чтением обратите внимание на разделитель внутри файла.
Чему равна размерность таблицы? Введите её в виде кортежа (число строк, число столбцов).

In [9]:
df.shape

(44744, 12)

### Задание 2.2
Выведите несколько первых (последних) строк таблицы, чтобы убедиться в том, что ваши данные не повреждены. Ознакомьтесь с признаками и их структурой.
В каком формате представлен порядок слов в признаке «Пол, возраст»?

* [x] <Пол , возраст , дата рождения>
* [ ] <Возраст , дата рождения, пол>
* [ ] <Пол , возраст>
* [ ] <Пол , дата рождения>

In [10]:
df['Пол, возраст'].head(5)

0      Мужчина ,  39 лет , родился 27 ноября 1979
1       Мужчина ,  60 лет , родился 20 марта 1959
2    Женщина ,  36 лет , родилась 12 августа 1982
3        Мужчина ,  38 лет , родился 25 июня 1980
4       Женщина ,  26 лет , родилась 3 марта 1993
Name: Пол, возраст, dtype: object

###  Задание 2.3
Выведите основную информацию о числе непустых значений в столбцах и их типах в таблице.

Какой тип данных имеют все столбцы? Введите его в поле для ответа.

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44744 entries, 0 to 44743
Data columns (total 12 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   Пол, возраст                     44744 non-null  object
 1   ЗП                               44744 non-null  object
 2   Ищет работу на должность:        44744 non-null  object
 3   Город, переезд, командировки     44744 non-null  object
 4   Занятость                        44744 non-null  object
 5   График                           44744 non-null  object
 6   Опыт работы                      44576 non-null  object
 7   Последнее/нынешнее место работы  44743 non-null  object
 8   Последняя/нынешняя должность     44742 non-null  object
 9   Образование и ВУЗ                44744 non-null  object
 10  Обновление резюме                44744 non-null  object
 11  Авто                             44744 non-null  object
dtypes: object(12)
memory usage: 4.1+

###  Задание 2.4

Обратите внимание на информацию о числе непустых значений.
Выберите все признаки, в которых есть пропуски в данных:
* [ ] Занятость
* [x] Последняя/нынешняя должность
* [ ] График
* [ ] Зарплата
* [x] Опыт работы
* [x] Последнее/нынешнее место работы


 Три колонки имеют пропуски формально по признаку того, что количество записей меньше, чем количество строк в таблице:
 * Опыт работы                      44576 / 44744
 * Последнее/нынешнее место работы  44743 / 44744
 * Последняя/нынешняя должность     44742 / 44744

 Кроме того, в колонках "Последняя/нынешняя должность" и "Последнее/нынешнее место работы" есть текстовые иденитификаторы, которые можно интерпретировать, как отсутствие записей (`-`, `..`, `/`, `.` и др.).

###  Задание 2.5
Выведите основную статистическую информацию о столбцах.
* Сколько уникальных значений содержится в столбце «Опыт работы»?
* Какая из искомых должностей самая распространённая? Введите название.



In [12]:
df["Опыт работы"].nunique()

44413

In [13]:
df["Ищет работу на должность:"].value_counts()

Ищет работу на должность:
Системный администратор                            3099
Инженер                                             966
Аналитик                                            809
Менеджер проектов                                   745
Руководитель проекта                                718
                                                   ... 
Email-маркетолог, Автор                               1
Специалист технической поддержки, IT специалист       1
Программист .NET junior                               1
Data Scientist, Data Analytics                        1
Инженер IT / Техн⁢ический специалист                  1
Name: count, Length: 14929, dtype: int64

# Преобразование данных

## Образование и ВУЗы

1. Начнем с простого - с признака **"Образование и ВУЗ"**. Его текущий формат это: **<Уровень образования год выпуска ВУЗ специальность...>**. Например:
* Высшее образование 2016 Московский авиационный институт (национальный исследовательский университет)...
* Неоконченное высшее образование 2000  Балтийская государственная академия рыбопромыслового флота…
Нас будет интересовать только уровень образования.

Создайте с помощью функции-преобразования новый признак **"Образование"**, который должен иметь 4 категории: "высшее", "неоконченное высшее", "среднее специальное" и "среднее".

Выполните преобразование, ответьте на контрольные вопросы и удалите признак "Образование и ВУЗ".

Совет: обратите внимание на структуру текста в столбце **"Образование и ВУЗ"**. Гарантируется, что текущий уровень образования соискателя всегда находится в первых 2ух слов и начинается с заглавной буквы. Воспользуйтесь этим.

*Совет: проверяйте полученные категории, например, с помощью метода unique()*


In [14]:
edu_year_r = r"\s*(\d{4})\s*"

def edu_mapping(edu_str):
  edu = edu_str.lower()
  if edu.startswith("среднее специальное"):
    return "среднее специальное"
  elif edu.startswith("среднее образование"):
    return "среднее"
  elif edu.startswith("неоконченное высшее"):
    return "неоконченное высшее"
  elif edu.startswith("высшее образование"):
    return "высшее"

df['Образование'] = df['Образование и ВУЗ'].str.split(edu_year_r, n=1, expand=True)[0].map(edu_mapping)

In [15]:
df['Образование'].unique()

array(['неоконченное высшее', 'высшее', 'среднее специальное', 'среднее'],
      dtype=object)

###  Задание 3.1
Сколько соискателей имеет средний уровень образования (школьное образование)?

In [16]:
df['Образование'].value_counts()

Образование
высшее                 33863
среднее специальное     5765
неоконченное высшее     4557
среднее                  559
Name: count, dtype: int64

In [17]:
df = df.drop(['Образование и ВУЗ'], axis="columns")

## Пол и возраст
Теперь нас интересует столбец **"Пол, возраст"**. Сейчас он представлен в формате **<Пол , возраст , дата рождения >**. Например:
* Мужчина , 39 лет , родился 27 ноября 1979
* Женщина , 21 год , родилась 13 января 2000
Как вы понимаете, нам необходимо выделить каждый параметр в отдельный столбец.

Создайте два новых признака **"Пол"** и **"Возраст"**. При этом важно учесть:
* Признак пола должен иметь 2 уникальных строковых значения: 'М' - мужчина, 'Ж' - женщина.
* Признак возраста должен быть представлен целыми числами.

Выполните преобразование, ответьте на контрольные вопросы и удалите признак **"Пол, возраст"** из таблицы.

*Совет: обратите внимание на структуру текста в столбце, в части на то, как разделены параметры пола, возраста и даты рождения между собой - символом ' , '.
Гарантируется, что структура одинакова для всех строк в таблице. Вы можете воспользоваться этим.*


In [18]:
gender_age = df["Пол, возраст"].str.split(",", expand=True)[[0, 1]]
df["Пол"] = gender_age[0].str.strip().map({"Мужчина": "М", "Женщина": "Ж"})
df["Возраст"] = gender_age[1].str.replace(r"[лет|года]", "", regex=True).str.strip().astype(int)

In [19]:
df = df.drop(["Пол, возраст"], axis="columns")

###  Задание 3.2
Сколько процентов женских резюме представлено в наших данных? Ответ округлите до сотых.

Чему равен средний возраст соискателей? Ответ округлите до десятых.

In [20]:
df["Пол"].value_counts(normalize=True).round(4)*100

Пол
М    80.93
Ж    19.07
Name: proportion, dtype: float64

In [21]:
df["Возраст"].mean().round(2)

np.float64(32.2)

## Опыт работы
Следующим этапом преобразуем признак **"Опыт работы"**. Его текущий формат - это: **<Опыт работы: n лет m месяцев, периоды работы в различных компаниях…>**.

Из столбца нам необходимо выделить общий опыт работы соискателя в месяцах, новый признак назовем "Опыт работы (месяц)"

Для начала обсудим условия решения задачи:
* Во-первых, в данном признаке есть пропуски. Условимся, что если мы встречаем пропуск, оставляем его как есть (функция-преобразование возвращает NaN)
* Во-вторых, в данном признаке есть скрытые пропуски. Для некоторых соискателей в столбце стоит значения "Не указано". Их тоже обозначим как NaN (функция-преобразование возвращает NaN)
* В-третьих, нас не интересует информация, которая описывается после указания опыта работы (периоды работы в различных компаниях)
* В-четвертых, у нас есть проблема: опыт работы может быть представлен только в годах или только месяцах. Например, можно встретить следующие варианты:
    * Опыт работы 3 года 2 месяца…
    * Опыт работы 4 года…
    * Опыт работы 11 месяцев…
    * Учитывайте эту особенность в вашем коде

Учитывайте эту особенность в вашем коде

В результате преобразования у вас должен получиться столбец, содержащий информацию о том, сколько месяцев проработал соискатель.
Выполните преобразование, ответьте на контрольные вопросы и удалите столбец **"Опыт работы"** из таблицы.


In [22]:
test_str = df["Опыт работы"].tolist()

In [23]:
def extract_exp_months(exp_str):
  if exp_str is None:
    return exp_str
  if exp_str == 'Не указано':
    return np.nan

  exp_r = r"^(Опыт работы) ((\d{1,3} (лет|года|год) \d{1,3} (месяцев|месяца|месяц))|\d{1,3} (месяцев|месяца|месяц)|(\d{1,3} (лет|года|год)))"
  years_r = r"(лет|года|год)"
  months_r = r"(месяцев|месяца|месяц)"
  year_month_r = r"(\d{1,3}) (лет|года|год) (\d{1,3}) (месяцев|месяца|месяц)"
  year_only_r = r"(\d{1,3}) (лет|года|год)"
  month_only_r = r"(\d{1,3}) (месяцев|месяца|месяц)"
  target_match = re.search(exp_r, str(exp_str))

  if target_match:
    matched_seq = target_match.group().replace("Опыт работы", "").strip()
    year_month_match = re.search(year_month_r, matched_seq)
    if year_month_match:
      y, m = int(year_month_match.group(1)), int(year_month_match.group(3))
      return y*12 + m

    year_only_match = re.search(year_only_r, matched_seq)
    if year_only_match:
      y = int(year_only_match.group(1))
      return y*12

    month_only_match = re.search(month_only_r, matched_seq)
    if month_only_match:
      m = int(month_only_match.group(1))
      return m

  else:
    return exp_str

In [24]:
df["Опыт работы (месяц)"] = df["Опыт работы"].apply(extract_exp_months)

###  Задание 3.3
Чему равен медианный опыт работы (в месяцах) в нашей таблице?

In [25]:
df["Опыт работы (месяц)"].median()

np.float64(100.0)

In [26]:
df = df.drop(["Опыт работы"], axis='columns')

## Город, Готовность к переезду, Готовность к командировкам

Хорошо идем! Следующий на очереди признак "Город, переезд, командировки". Информация в нем представлена в следующем виде: **<Город , (метро) , готовность к переезду (города для переезда) , готовность к командировкам>**. В скобках указаны необязательные параметры строки. Например, можно встретить следующие варианты:

* Москва , не готов к переезду , готов к командировкам
* Москва , м. Беломорская , не готов к переезду, не готов к командировкам
* Воронеж , готов к переезду (Сочи, Москва, Санкт-Петербург) , готов к командировкам

Создадим отдельные признаки **"Город"**, **"Готовность к переезду"**, **"Готовность к командировкам"**. При этом важно учесть:

* Признак **"Город"** должен содержать только 4 категории: "Москва", "Санкт-Петербург" и "город-миллионник" (их список ниже), остальные обозначьте как "другие".

    Список городов-миллионников:
    
   <code>million_cities = ['Новосибирск', 'Екатеринбург','Нижний Новгород','Казань', 'Челябинск','Омск', 'Самара', 'Ростов-на-Дону', 'Уфа', 'Красноярск', 'Пермь', 'Воронеж','Волгоград']
    </code>
    Инфорация о метро, рядом с которым проживает соискатель нас не интересует.
* Признак **"Готовность к переезду"** должен иметь два возможных варианта: True или False. Обратите внимание, что возможны несколько вариантов описания готовности к переезду в признаке "Город, переезд, командировки". Например:
    * … , готов к переезду , …
    * … , не готова к переезду , …
    * … , готова к переезду (Москва, Санкт-Петербург, Ростов-на-Дону)
    * … , хочу переехать (США) , …
    
    Нас интересует только сам факт возможности или желания переезда.
* Признак **"Готовность к командировкам"** должен иметь два возможных варианта: True или False. Обратите внимание, что возможны несколько вариантов описания готовности к командировкам в признаке "Город, переезд, командировки". Например:
    * … , готов к командировкам , …
    * … , готова к редким командировкам , …
    * … , не готов к командировкам , …
    
    Нас интересует только сам факт готовности к командировке.
    
    Еще один важный факт: при выгрузки данных у некоторых соискателей "потерялась" информация о готовности к командировкам. Давайте по умолчанию будем считать, что такие соискатели не готовы к командировкам.
    
Выполните преобразования и удалите столбец **"Город, переезд, командировки"** из таблицы.

*Совет: обратите внимание на то, что структура текста может меняться в зависимости от указания ближайшего метро. Учите это, если будете использовать порядок слов в своей программе.*


In [27]:
def not_ready_for_trips(expr):
  return True if re.search(r"(не готов(а)? к командировкам)$", expr) else False

def not_ready_for_moving(expr):
  return True if re.search(r"не готов(а)? к переезду", expr) else False

def map_city(city):
  million_cities = ['Новосибирск', 'Екатеринбург','Нижний Новгород','Казань', 'Челябинск','Омск', 'Самара', 'Ростов-на-Дону', 'Уфа', 'Красноярск', 'Пермь', 'Воронеж','Волгоград']
  if city in ["Москва", "Санкт-Петербург"]:
    return city
  elif city in million_cities:
    return "город-миллионник"
  else:
    return "другие"

split_1 = df['Город, переезд, командировки'].str.rsplit(",", n=1, expand=True)
trips = split_1[1]
split_2 = split_1[0].str.split(",", n=1, expand=True)
city = split_2[0].str.strip()
moving = split_2[1]

df['Готовность к командировкам'] = ~trips.apply(not_ready_for_trips)
df['Готовность к переезду'] = ~moving.apply(not_ready_for_moving)
df['Город'] = city.map(map_city)

###  Задание 3.4

* Сколько процентов соискателей живут в Санкт-Петербурге? Ответ округлите до целого.

* Сколько процентов соискателей готовы одновременно и к переездам, и к командировкам? Ответ округлите до целого.

In [28]:
df['Город'].value_counts(normalize=True).mul(100).round(0)

Город
Москва              37.0
другие              35.0
город-миллионник    16.0
Санкт-Петербург     11.0
Name: proportion, dtype: float64

In [29]:
(df['Готовность к командировкам'] & df['Готовность к переезду']).value_counts(normalize=True).mul(100).round(1)

False    68.0
True     32.0
Name: proportion, dtype: float64

In [30]:
df = df.drop(['Город, переезд, командировки'], axis="columns")
df.columns

Index(['ЗП', 'Ищет работу на должность:', 'Занятость', 'График',
       'Последнее/нынешнее место работы', 'Последняя/нынешняя должность',
       'Обновление резюме', 'Авто', 'Образование', 'Пол', 'Возраст',
       'Опыт работы (месяц)', 'Готовность к командировкам',
       'Готовность к переезду', 'Город'],
      dtype='object')

## Занятость и График
Рассмотрим поближе признаки **"Занятость"** и **"График"**. Сейчас признаки представляют собой набор категорий желаемой занятости (полная занятость, частичная занятость, проектная работа, волонтерство, стажировка) и желаемого графика работы (полный день, сменный график, гибкий график, удаленная работа, вахтовый метод).
На сайте hh.ru соискатель может указывать различные комбинации данных категорий, например:
* полная занятость, частичная занятость
* частичная занятость, проектная работа, волонтерство
* полный день, удаленная работа
* вахтовый метод, гибкий график, удаленная работа, полная занятость

Такой вариант признаков имеет множество различных комбинаций, а значит множество уникальных значений, что мешает анализу. Нужно это исправить!

Давайте создадим признаки-мигалки для каждой категории: если категория присутствует в списке желаемых соискателем, то в столбце на месте строки рассматриваемого соискателя ставится True, иначе - False.

Такой метод преобразования категориальных признаков называется One Hot Encoding и его схема представлена на рисунке ниже:
<img src=https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/ohe.jpg>
Выполните данное преобразование для признаков "Занятость" и "График", ответьте на контрольные вопросы, после чего удалите их из таблицы

In [31]:
uniq_occupation = list(set(chain(*[[uniq.strip() for uniq in occ.split(",")] for occ in df['Занятость'].unique()])))
for occ in uniq_occupation:
  df[occ] = df['Занятость'].str.contains(occ)

In [32]:
df[uniq_occupation].value_counts()

частичная занятость  проектная работа  полная занятость  стажировка  волонтерство
False                False             True              False       False           30026
True                 False             True              False       False            4835
                     True              True              False       False            4762
                                                         True        False            1079
False                True              True              False       False             985
True                 False             False             False       False             630
                                       True              True        False             613
                     True              False             False       False             564
False                False             True              True        False             444
True                 True              True              True        True              341
        

In [33]:
uniq_schedule = list(set(chain(*[[uniq.strip() for uniq in sched.split(",")] for sched in df['График'].unique()])))
for sched in uniq_schedule:
  df[sched] = df['График'].str.contains(sched)

In [34]:
df[uniq_schedule].value_counts()

сменный график  полный день  удаленная работа  гибкий график  вахтовый метод
False           True         False             False          False             22727
True            True         True              True           False              4791
False           True         True              True           False              3198
True            True         False             True           False              2275
False           True         True              False          False              1935
True            True         False             False          False              1834
                             True              True           True               1829
False           True         False             True           False              1450
                False        True              False          False              1108
                                               True           False               782
True            True         True              False          F

###  Задание 3.5
* Сколько людей ищут проектную работу и волонтерство (в обоих столбцах стоит True)?
* Сколько людей хотят работать вахтовым методом и с гибким графиком (в обоих столбцах стоит True)?


In [35]:
(df['проектная работа'] & df['волонтерство']).sum()

np.int64(436)

In [36]:
(df['вахтовый метод'] & df['гибкий график']).sum()

np.int64(2311)

## ЗП

Наконец, мы добрались до самого главного и самого важного - признака заработной платы **"ЗП"**.
В чем наша беда? В том, что помимо желаемой заработной платы соискатель указывает валюту, в которой он бы хотел ее получать, например:
* 30000 руб.
* 50000 грн.
* 550 USD

Нам бы хотелось видеть заработную плату в единой валюте, например, в рублях. Возникает вопрос, а где взять курс валют по отношению к рублю?

На самом деле язык Python имеет в арсенале огромное количество возможностей получения данной информации, от обращения к API Центробанка, до использования специальных библиотек, например pycbrf. Однако, это не тема нашего проекта.

Поэтому мы пойдем в лоб: обратимся к специальным интернет-ресурсам для получения данных о курсе в виде текстовых файлов. Например, MDF.RU, данный ресурс позволяет удобно экспортировать данные о курсах различных валют и акций за указанные периоды в виде csv файлов. Мы уже сделали выгрузку курсов валют, которые встречаются в наших данных за период с 29.12.2017 по 05.12.2019. Скачать ее вы можете **на платформе**

Создайте новый DataFrame из полученного файла. В полученной таблице нас будут интересовать столбцы:
* "currency" - наименование валюты в ISO кодировке,
* "date" - дата,
* "proportion" - пропорция,
* "close" - цена закрытия (последний зафиксированный курс валюты на указанный день).


Перед вами таблица соответствия наименований иностранных валют в наших данных и их общепринятых сокращений, которые представлены в нашем файле с курсами валют. Пропорция - это число, за сколько единиц валюты указан курс в таблице с курсами. Например, для казахстанского тенге курс на 20.08.2019 составляет 17.197 руб. за 100 тенге, тогда итоговый курс равен - 17.197 / 100 = 0.17197 руб за 1 тенге.
Воспользуйтесь этой информацией в ваших преобразованиях.

<img src=https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/table.jpg>


Осталось только понять, откуда брать дату, по которой определяется курс? А вот же она - в признаке **"Обновление резюме"**, в нем содержится дата и время, когда соискатель выложил текущий вариант своего резюме. Нас интересует только дата, по ней бы и будем сопоставлять курсы валют.

Теперь у нас есть вся необходимая информация для того, чтобы создать признак "ЗП (руб)" - заработная плата в рублях.

После ответа на контрольные вопросы удалите исходный столбец заработной платы "ЗП" и все промежуточные столбцы, если вы их создавали.

Итак, давайте обсудим возможный алгоритм преобразования:
1. Перевести признак "Обновление резюме" из таблицы с резюме в формат datetime и достать из него дату. В тот же формат привести признак "date" из таблицы с валютами.
2. Выделить из столбца "ЗП" сумму желаемой заработной платы и наименование валюты, в которой она исчисляется. Наименование валюты перевести в стандарт ISO согласно с таблицей выше.
3. Присоединить к таблице с резюме таблицу с курсами по столбцам с датой и названием валюты (подумайте, какой тип объединения надо выбрать, чтобы в таблице с резюме сохранились данные о заработной плате, изначально представленной в рублях). Значение close для рубля заполнить единицей 1 (курс рубля самого к себе)
4. Умножить сумму желаемой заработной платы на присоединенный курс валюты (close) и разделить на пропорцию (обратите внимание на пропуски после объединения в этих столбцах), результат занести в новый столбец "ЗП (руб)".


In [37]:
ExchangeRates = pd.read_csv("https://lms-cdn.skillfactory.ru/assets/courseware/v1/15abf80f45a2f3e93c3274101b451c67/asset-v1:skillfactory+MIFIML-1sem+2025+type@asset+block/ExchangeRates.zip")
ExchangeRates.head()

currency per      date   time    close  vol  proportion
0      USD   D  29/12/17  00:00  57.6291    0           1
1      USD   D  30/12/17  00:00  57.6002    0           1
2      USD   D  31/12/17  00:00  57.6002    0           1
3      USD   D  01/01/18  00:00  57.6002    0           1
4      USD   D  02/01/18  00:00  57.6002    0           1

In [38]:
ExchangeRates['date'] = pd.to_datetime(ExchangeRates['date'], format="%d/%m/%y").dt.date
ExchangeRates['date'].head()

0    2017-12-29
1    2017-12-30
2    2017-12-31
3    2018-01-01
4    2018-01-02
Name: date, dtype: object

In [39]:
df['Дата обновления резюме'] = pd.to_datetime(df['Обновление резюме'], dayfirst=True).dt.date
df['Дата обновления резюме'].head()

0    2019-04-16
1    2019-04-12
2    2019-04-16
3    2019-04-08
4    2019-04-22
Name: Дата обновления резюме, dtype: object

In [40]:
curr_map = {"руб.": "RUB",
            'бел.руб.': "BYN",
            'KZT': "KZT",
            'EUR': "EUR",
            'USD': "USD",
            'грн.': "UAH",
            'сум': "UZS",
            'KGS': "KGS",
            'AZN': "AZN"}
salary = df['ЗП'].str.split(" ", expand=True)
salary.columns = ['Зарплата', 'Валюта']
salary['Зарплата'] = salary['Зарплата'].astype(np.float64)
salary['Валюта'] = salary['Валюта'].map(curr_map)
salary

Зарплата Валюта
0       29000.0    RUB
1       40000.0    RUB
2       20000.0    RUB
3      100000.0    RUB
4      140000.0    RUB
...         ...    ...
44739   50000.0    RUB
44740   39000.0    RUB
44741   40000.0    RUB
44742   20000.0    RUB
44743  120000.0    RUB

[44744 rows x 2 columns]

In [41]:
df = pd.concat([df, salary], axis=1)
df.head()

ЗП                          Ищет работу на должность:  \
0   29000 руб.                            Системный администратор   
1   40000 руб.                               Технический писатель   
2   20000 руб.                                           Оператор   
3  100000 руб.  Веб-разработчик (HTML / CSS / JS / PHP / базы ...   
4  140000 руб.                  Региональный менеджер по продажам   

                                           Занятость  \
0  частичная занятость, проектная работа, полная ...   
1  частичная занятость, проектная работа, полная ...   
2                                   полная занятость   
3  частичная занятость, проектная работа, полная ...   
4                                   полная занятость   

                                              График  \
0  гибкий график, полный день, сменный график, ва...   
1  гибкий график, полный день, сменный график, уд...   
2                                        полный день   
3                    гибкий график, удаленная работа   
4                                        полный день   

  Последнее/нынешнее место работы  \
0         МАОУ "СОШ № 1 г.Немана"   
1    Временный трудовой коллектив   
2                    ПАО Сбербанк   
3                        OpenSoft   
4                        Мармелад   

                       Последняя/нынешняя должность Обновление резюме  \
0                           Системный администратор  16.04.2019 15:59   
1  Менеджер проекта, Аналитик, Технический писатель  12.04.2019 08:42   
2                               Кассир-операционист  16.04.2019 08:35   
3                               Инженер-программист  08.04.2019 14:23   
4                              Менеджер по продажам  22.04.2019 10:32   

                             Авто          Образование Пол  ...  стажировка  \
0  Имеется собственный автомобиль  неоконченное высшее   М  ...       False   
1                      Не указано               высшее   М  ...       False   
2                      Не указано  среднее специальное   Ж  ...       False   
3                      Не указано               высшее   М  ...       False   
4                      Не указано               высшее   Ж  ...       False   

   волонтерство  сменный график  полный день удаленная работа  гибкий график  \
0         False            True         True             True           True   
1         False            True         True             True           True   
2         False           False         True            False          False   
3         False           False        False             True           True   
4         False           False         True            False          False   

   вахтовый метод  Дата обновления резюме  Зарплата  Валюта  
0            True              2019-04-16   29000.0     RUB  
1           False              2019-04-12   40000.0     RUB  
2           False              2019-04-16   20000.0     RUB  
3           False              2019-04-08  100000.0     RUB  
4           False              2019-04-22  140000.0     RUB  

[5 rows x 28 columns]

In [42]:
df_with_rates = pd.merge(
    df,
    ExchangeRates[['date', 'currency', 'close', 'proportion']],
    how='left',
    left_on=['Дата обновления резюме', 'Валюта'],
    right_on=['date', 'currency'])

In [43]:
df_with_rates.loc[df_with_rates['Валюта'] == 'RUB', 'close'] = 1.0
df_with_rates.loc[df_with_rates['Валюта'] == 'RUB', 'proportion'] = 1.0

In [44]:
df_with_rates["ЗП (руб)"] = (df_with_rates['Зарплата'] * df_with_rates['close']) / df_with_rates['proportion']

###  Задание 3.6

Для выполнения этого задания у вас будет 5 попыток. После того как вы дадите правильный ответ ИЛИ истратите все попытки, вы можете нажать на кнопку «Показать ответ»: в конце решения вы увидите ссылку на эталонный ответ.

Чему равна желаемая медианная заработная плата соискателей в нашей таблице (в рублях)?

Ответ приведите в тысячах рублей и округлите до целого. Например, если у вас получилась медианная з/п 42 127, в качестве ответа укажите 42.

In [45]:
round(df_with_rates["ЗП (руб)"].median() / 1e3, 0)

np.float64(59.0)

In [46]:
df_with_rates = df_with_rates.drop(['ЗП', 'Валюта', 'proportion', 'close', 'Зарплата', 'date', 'currency', 'Дата обновления резюме'], axis='columns')

# Исследование зависимостей в данных

## Возраст
Постройте распределение признака **"Возраст"**. Опишите распределение, отвечая на следующие вопросы: чему равна мода распределения, каковы предельные значения признака, в каком примерном интервале находится возраст большинства соискателей? Есть ли аномалии для признака возраста, какие значения вы бы причислили к их числу?
*Совет: постройте гистограмму и коробчатую диаграмму рядом.*

In [47]:
# fig = px.histogram(df_with_rates, x="Возраст", title="Гистограмма по признаку 'Возраст'")
# fig.update_layout(xaxis_title="Возраст", yaxis_title="Количество наблюдений")
# fig.show()
# fig.write_html("plotly_graphs/01_histogram_age.html")
IFrame(src='plotly_graphs/01_histogram_age.html', width=1000, height=600)

In [48]:
# fig = px.box(df_with_rates, y="Возраст", title="Коробчатая диаграмма по признаку 'Возраст'")
# fig.update_layout(yaxis_title="Возраст")
# fig.show()
# fig.write_html("plotly_graphs/02_boxplot_age.html")
IFrame(src='plotly_graphs/02_boxplot_age.html', width=1000, height=600)

In [49]:
(df_with_rates['Возраст'] > 65).sum()

np.int64(45)

**Вопрос**: Чему равна мода распределения?

**Ответ**: Мода распределения равна 30 годам с 2834 значениями.

----

**Вопрос**: Каковы предельные значения признака?

**Ответ**: Предельные значения признака "Возраст" 14 и 100 лет.

----

**Вопрос**: В каком примерном интервале находится возраст большинства соискателей?

**Ответ**: Возраст большинства соискателей находится в диапазоне от 27 лет (Q1) до 36 лет (Q3).

----

**Вопрос**: Есть ли аномалии для признака возраста, какие значения вы бы причислили к их числу?

**Ответ**: Есть одна аномалия для признака возраст - 100 лет. Если рассматривать более формально, то предельными значениями можно взять  пенсионный возраст мужчин в 2025г. (65 лет) и самый ранный возраст - 14 лет при наличии согласия родителей.
В таком подходе, все возраста меньше 14 и свыше 65 лет можно отнести к аномалиям. Меньше 14 лет значений не наблюдается. Свыше 65 лет есть 45 значений.

## Опыт работы (месяц)

Постройте распределение признака **"Опыт работы (месяц)"**. Опишите данное распределение, отвечая на следующие вопросы: чему равна мода распределения, каковы предельные значения признака, в каком примерном интервале находится опыт работы большинства соискателей? Есть ли аномалии для признака опыта работы, какие значения вы бы причислили к их числу?
*Совет: постройте гистограмму и коробчатую диаграмму рядом.*

In [50]:
# fig = px.histogram(df_with_rates, x="Опыт работы (месяц)", title="Гистограмма по признаку 'Опыт работы (месяц)'")
# fig.update_layout(xaxis_title="Опыт работы (месяц)", yaxis_title="Количество наблюдений")
# fig.show()
# fig.write_html("plotly_graphs/03_hist_experience.html")
IFrame(src='plotly_graphs/03_hist_experience.html', width=1000, height=600)

In [51]:
df_with_rates["Опыт работы (месяц)"].mode()

0    81.0
Name: Опыт работы (месяц), dtype: float64

In [52]:
(df_with_rates["Опыт работы (месяц)"] == 81).sum()

np.int64(313)

In [53]:
# fig = px.box(df_with_rates, y="Опыт работы (месяц)", title="Коробчатая диаграмма по признаку 'Опыт работы (месяц)'")
# fig.update_layout(yaxis_title="Опыт работы (месяц)")
# fig.show()
# fig.write_html("plotly_graphs/04_boxplot_experience.html")
IFrame(src='plotly_graphs/04_boxplot_experience.html', width=1000, height=600)

In [54]:
(df_with_rates["Опыт работы (месяц)"] > 612).sum()

np.int64(4)

**Вопрос**: Чему равна мода распределения?

**Ответ**: Мода распределения находится в диапазоне 80-84 месяцев в количестве 1414. Если точнее, то 81 месяц в количестве 313 значений.

----

**Вопрос**: Каковы предельные значения признака?

**Ответ**: Предельные значения признака "Опыт работы" составляют 1 и 1188 месяцев.

----

**Вопрос**: В каком примерном интервале находится опыт работы большинства соискателей?

**Ответ**: Опыт работы большинства соискателей находится в диапазоне от  57 (Q1) до 154 (Q3).

----

**Вопрос**: Есть ли аномалии для признака "Опыт работы", какие значения вы бы причислили к их числу?

**Ответ**: Есть одна аномалия для признака опыт работы - 1188 месяцев (99 лет трудового стажа!). Если рассматривать более формально, то предельными значениями можно взять  трудовой стаж 612 месяцев ($(65 \text{ лет} - 14 \text{ лет}) * 12 \text{ месяцев} = 612 \text{ месяцев}$).
В таком подходе, весь стаж свыше 612 месяцев можно отнести к аномалиям. Свыше 612 месяцев есть 4 значения.

## ЗП (руб)
Постройте распределение признака **"ЗП (руб)"**. Опишите данное распределение, отвечая на следующие вопросы: каковы предельные значения признака, в каком примерном интервале находится заработная плата большинства соискателей? Есть ли аномалии для признака возраста? Обратите внимание на гигантские размеры желаемой заработной платы.
*Совет: постройте гистограмму и коробчатую диаграмму рядом.*


In [55]:
# fig = px.histogram(df_with_rates, x="ЗП (руб)", title="Гистограмма по признаку 'ЗП (руб)'")
# fig.update_layout(xaxis_title="ЗП (руб)", yaxis_title="Количество наблюдений")
# # fig.show()
# fig.write_html("plotly_graphs/05_hist_salary.html")
IFrame(src='plotly_graphs/05_hist_salary.html', width=1000, height=600)

In [56]:
df_with_rates["ЗП (руб)"].max()

np.float64(24304876.0)

In [57]:
# fig = px.box(df_with_rates, y="ЗП (руб)", title="Коробчатая диаграмма по признаку 'ЗП (руб)'")
# fig.update_layout(yaxis_title="ЗП (руб)")
# fig.show()
# fig.write_html("plotly_graphs/06_boxplot_salary.html")
IFrame(src='plotly_graphs/06_boxplot_salary.html', width=1000, height=600)

In [58]:
(df_with_rates["ЗП (руб)"] > 1e6).sum()

np.int64(5)

In [59]:
(df_with_rates["ЗП (руб)"] > 500e3).sum()

np.int64(32)

**Вопрос**: Каковы предельные значения признака?

**Ответ**: Предельные значения признака "Опыт работы" составляют 1 руб. и  24 304 876 рублей.

----

**Вопрос**: В каком примерном интервале находится заработная плата большинства соискателей?

**Ответ**: Опыт работы большинства соискателей находится в диапазоне от 37 тысяч рублей  (Q1) до 95 тысяч рублей (Q3).

----

**Вопрос**: Есть ли аномалии для признака "заработная плата", какие значения вы бы причислили к их числу?

**Ответ**: Есть аномалии для признака заработная плата. За такой порог можно взять зарплаты свыше 500 тысяч рублей. Одна из таких зарплат составляет 24 304 876 рублей. Всего зарплат свыше 500 тысяч рублей - 32 наблюдения.

## Зависимость **медианной** желаемой заработной платы ЗП (руб)
Постройте диаграмму, которая показывает зависимость **медианной** желаемой заработной платы (**"ЗП (руб)"**) от уровня образования (**"Образование"**).
Используйте для диаграммы данные о резюме, где желаемая заработная плата меньше 1 млн рублей.
*Сделайте выводы по представленной диаграмме: для каких уровней образования наблюдаются наибольшие и наименьшие уровни желаемой заработной платы? Как вы считаете, важен ли признак уровня образования при прогнозировании заработной платы?*

In [60]:
salaries_by_edu = (
    df_with_rates[df_with_rates['ЗП (руб)'] < 1e6]
    .groupby("Образование", as_index=False)['ЗП (руб)']
    .median()
    .sort_values('ЗП (руб)', ascending=True))

In [61]:
# fig = px.line(salaries_by_edu, x="Образование", y='ЗП (руб)', title="Зависимость 'ЗП (руб)' от образования")
# fig.update_layout(yaxis_title="Медианная ЗП (руб)")
# fig.show()
# fig.write_html("plotly_graphs/07_salary_vs_edu.html")
IFrame(src='plotly_graphs/07_salary_vs_edu.html', width=1000, height=600)

**Вопрос**: для каких уровней образования наблюдаются наибольшие и наименьшие уровни желаемой заработной платы?

**Ответ**:
* Наибольшие уровни желаемой заработной платы наблюдаются для высшего и неоконченного высшего образования.
* Наименьшие уровни желаемой заработной платы наблюдаются для среднего и среднеспециального образования.

---

**Вопрос**: Как вы считаете, важен ли признак уровня образования при прогнозировании заработной платы?

**Ответ**: Считаю, что важен. Люди с высшим образованиям склонны желать более высокой оплаты. Более квалифицированный труд, обычно, оплачивается выше.

## Распределение желаемой заработной платы (**"ЗП (руб)"**) в зависимости от города (**"Город"**)
Постройте диаграмму, которая показывает распределение желаемой заработной платы (**"ЗП (руб)"**) в зависимости от города (**"Город"**). Используйте для диаграммы данные о резюме, где желая заработная плата меньше 1 млн рублей.
*Сделайте выводы по полученной диаграмме: как соотносятся медианные уровни желаемой заработной платы и их размах в городах? Как вы считаете, важен ли признак города при прогнозировании заработной платы?*

In [62]:
salaries_by_city = (
    df_with_rates[df_with_rates['ЗП (руб)'] < 1e6]
    .groupby("Город", as_index=False)['ЗП (руб)']
    .agg(["median", "min", "max"])
    .assign(range = lambda _df: _df['max'] - _df['min'])
    .sort_values('median', ascending=True))
salaries_by_city

Город   median  min       max     range
3            другие  40000.0  1.0  923983.0  923982.0
2  город-миллионник  40000.0  1.0  511681.6  511680.6
1   Санкт-Петербург  60000.0  1.0  645171.0  645170.0
0            Москва  85000.0  1.0  900000.0  899999.0

In [63]:
# fig = px.line(salaries_by_city, x="Город", y='median', title="Зависимость ЗП (руб) от города")
# fig.update_layout(yaxis_title="Медианная ЗП (руб)")
# fig.show()
# fig.write_html("plotly_graphs/08_salary_vs_city.html")
IFrame(src='plotly_graphs/08_salary_vs_city.html', width=1000, height=600)

In [64]:
(df_with_rates[df_with_rates['ЗП (руб)'] < 1e6]
 .groupby("Город", as_index=False)['ЗП (руб)']
 .agg("max")
 .sort_values('ЗП (руб)', ascending=True))

Город  ЗП (руб)
2  город-миллионник  511681.6
1   Санкт-Петербург  645171.0
0            Москва  900000.0
3            другие  923983.0

In [65]:
from scipy.stats import iqr
iqr_salary_by_city = (
    df_with_rates[df_with_rates['ЗП (руб)'] < 1e6]
    .groupby("Город", as_index=False)['ЗП (руб)']
    .agg(iqr)
    .sort_values('ЗП (руб)', ascending=True))
iqr_salary_by_city

Город  ЗП (руб)
3            другие  30000.00
2  город-миллионник  30000.00
1   Санкт-Петербург  40227.86
0            Москва  90000.00




**Вопрос**: Как соотносятся медианные уровни желаемой заработной платы и их размах в городах?

**Ответ**:
* Наибольшие уровни желаемой заработной платы наблюдаются в Москве и Санкт-Петербурге. Причём медиана в Санкт-Петербурге в 1,5 раза выше, чем в городах-милионниках.
* Соответственно, чем крупнее город, тем выше медианная зарплата.
* Из-за того, что минимальные значения составляют 1 рубль, то размах от минимума до максима, примерно, равен максимуму. Это не очень информативно. Межквартильный размах более информативен. Для IQR тенденция та же - чем крупнее город, тем выше размах: для Москвы он составляет 90 000 рублей, а для малых городов - 30 000 руб.

---

**Вопрос**: Как вы считаете, важен ли признак города при прогнозировании заработной платы?

**Ответ**: Считаю, что важен. Люди проживающие в крупных городах склонны желать более высокой оплаты. Проживание в более крупных городах, обычно, стоит дороже.

## Зависимость медианной заработной платы ("ЗП (руб)") от признаков "Готовность к переезду" и "Готовность к командировкам"
Постройте **многоуровневую столбчатую диаграмму**, которая показывает зависимость медианной заработной платы (**"ЗП (руб)"**) от признаков **"Готовность к переезду"** и **"Готовность к командировкам"**. Проанализируйте график, сравнив уровень заработной платы в категориях.

In [66]:
salaries_by_moving_and_trips = (
    df_with_rates[df_with_rates['ЗП (руб)'] < 1e6]
    .groupby(['Готовность к командировкам', 'Готовность к переезду'], as_index=False)['ЗП (руб)']
    .median()
    .sort_values('ЗП (руб)', ascending=True))

salaries_by_moving_and_trips['Готовность к командировкам'] = (
    salaries_by_moving_and_trips['Готовность к командировкам']
    .map({True: "Готов(а) к командировкам", False: "Не готов(а) к командировкам"}))

salaries_by_moving_and_trips['Готовность к переезду'] = (
    salaries_by_moving_and_trips['Готовность к переезду']
    .map({True: "Готов(а) к переезду", False: "Не готов(а) к переезду"}))

salaries_by_moving_and_trips

Готовность к командировкам   Готовность к переезду  ЗП (руб)
0  Не готов(а) к командировкам  Не готов(а) к переезду   40000.0
1  Не готов(а) к командировкам     Готов(а) к переезду   50000.0
2     Готов(а) к командировкам  Не готов(а) к переезду   60000.0
3     Готов(а) к командировкам     Готов(а) к переезду   67000.0

In [67]:
# fig = px.bar(salaries_by_moving_and_trips,
#              x='Готовность к командировкам',
#              y='ЗП (руб)',
#              color="Готовность к переезду",
#              title="Зависимость медианной заработной платы от признаков готовности к переезду и командировкам",
#              barmode='group')
# fig.update_layout(yaxis_title="Медианная ЗП (руб)")
# fig.show()
fig.write_html("plotly_graphs/09_salary_vs_mobility.html")
IFrame(src='plotly_graphs/09_salary_vs_mobility.html', width=1000, height=600)

NameError: name 'fig' is not defined

На графике зависимости медианной зарплаты в рублях от готовности к переезду и командировкам видно, что уровень медианной зарплаты выше у тех кандидатов, которые более мобильны: готовы к командировкам и переезду.

Это можно объяснить тем, что командировки связаны с повышенным стрессом от поездок, а переезды связаны с более высокими расходами на переезд в новое место.

## Зависимость **медианной** желаемой заработной платы от возраста (**"Возраст"**) и образования (**"Образование"**)
Постройте сводную таблицу, иллюстрирующую зависимость **медианной** желаемой заработной платы от возраста (**"Возраст"**) и образования (**"Образование"**). На полученной сводной таблице постройте **тепловую карту**. Проанализируйте тепловую карту, сравнив показатели внутри групп.

In [ ]:
salaries_by_age_edu = (
    df_with_rates[df_with_rates['ЗП (руб)'] < 1e6]
    .assign(age_grp = lambda _df: pd.cut(_df['Возраст'], np.arange(10, 105, 5)))
    .groupby(['age_grp', 'Образование'], as_index=False, observed=True)['ЗП (руб)']
    .median()
    .pivot(index='Образование', columns='age_grp', values='ЗП (руб)')
    .rename_axis(None)
    )
salaries_by_age_edu.columns = [f"({interval.left}-{interval.right}]" for interval in salaries_by_age_edu.columns]
salaries_by_age_edu

In [ ]:
# fig = px.imshow(salaries_by_age_edu, text_auto=True, title="Зависимость медианной желаемой заработной платы от возраста и образования")
# fig.update_layout(xaxis_title="Возраст", yaxis_title="Образование")
# fig.show()
# fig.write_html("plotly_graphs/10_salary_vs_age_and_edu.html")
IFrame(src='plotly_graphs/10_salary_vs_age_and_edu.html', width=1000, height=600)

In [ ]:
df_with_rates[(df_with_rates['Возраст'] == 14) & (df_with_rates['Образование'] == "высшее")][['Ищет работу на должность:',
                                                                                              'Последнее/нынешнее место работы',
                                                                                              'Последняя/нынешняя должность',
                                                                                              'Возраст',
                                                                                              'ЗП (руб)']]

На тепловой карте можно заметить основные тенденции и аномалии.

Начнём с аномалий:
- В возрастном диапазоне (10-15] медианная зарплата для высшего образования составляет 60 тысяч рублей. Обычно, у 15-летних подростков ещё нет высшего образования. В данном случае, 14-летний Ведущий инженер-программист из КБ Аметист, Концерн АГАТ ищет работу на ту же должность с ЗП 60000.
- В возрастном диапазоне (65-70] медианная зарплата для среднеспециального образования составляет 120 тысяч рублей.
- Видим знакомого 100-летнего специалиста со среднеспециальным образованием, который ищет зарплату 60 тысяч рублей.

Обратим внимание на тенденции:
- Самые высокие зарплатные ожидания в группе с высшим образованием от 30 до 55 лет. В группе свыше пенсионного возраста 65-75 лет зарплатные ожидания повышаются  до 72,5-75 тысяч рублей.
- Для группы с неоконченным высшим образованием самые высокие зарплатные ожидания в возрасте от 40 до 50 лет и в возрасте 55-60 лет.

## Зависимость опыта работы от возраста
Постройте **диаграмму рассеяния**, показывающую зависимость опыта работы
(**"Опыт работы (месяц)"**) от возраста (**"Возраст"**). Опыт работы переведите из месяцев в года, чтобы признаки были в едином масштабе. Постройте на графике дополнительно прямую, проходящую через точки (0, 0) и (100, 100). Данная прямая соответствует значениям, когда опыт работы равен возрасту человека. Точки, лежащие на этой прямой и выше нее - аномалии в наших данных (опыт работы больше либо равен возрасту соискателя)

In [ ]:
# fig = px.scatter(df_with_rates,
#                  x="Возраст",
#                  y="Опыт работы (месяц)",
#                  title="Зависимость опыта работы от возраста")
# fig.add_shape(type="line", x0=0, y0=0, x1=100, y1=100, line_color="grey")
# fig.add_shape(type="line", x0=0, y0=0, x1=100, y1=1200, line_color="red")
# fig.add_shape(type="line", x0=14, y0=0, x1=100, y1=1200, line_color="blue")
# fig.update_layout(xaxis_title="Возраст", yaxis_title="Опыт работы (месяц)")
# fig.show()
# fig.write_html("plotly_graphs/11_age_vs_exp.html")
IFrame(src='plotly_graphs/11_age_vs_exp.html', width=1000, height=600)

Согласно условию задания добавлениа линия $A_1(0, 0) - A_2(100, 100)$ в сером цвете. Но т.к. опыт работы измерен в месяцах, то для отражения смысла (опыт работы больше либо равен возрасту соискателя), нужно, чтобы точка $A_2$ находилась в координатах $(100, 1200)$ - добавлена линия в красном цвете между точками $B_1(0, 0) - B_2(100, 1200)$.

С учётом правовых аспектов (работать можно с 14 лет), то добавлена линия в синем цвете между точками $C_1(14, 0) - C_2(100, 1200)$, которая более точно отсекает аномалии.

## **Дополнительные баллы**

Для получения 2 дополнительных баллов по разведывательному анализу постройте еще два любых содержательных графика или диаграммы, которые помогут проиллюстрировать влияние признаков/взаимосвязь между признаками/распределения признаков. Приведите выводы по ним. Желательно, чтобы в анализе участвовали признаки, которые мы создавали ранее в разделе "Преобразование данных".


### Зависимость зарплаты от пола и возраста

In [ ]:
salaries_by_age_gender = (
    df_with_rates[df_with_rates['ЗП (руб)'] < 1e6]
    .assign(age_grp = lambda _df: pd.cut(_df['Возраст'], np.arange(10, 105, 5)))
    .groupby(['age_grp', 'Пол'], as_index=False, observed=True)['ЗП (руб)']
    .median()
    .pivot(index='Пол', columns='age_grp', values='ЗП (руб)')
    .rename_axis(None)
    )
salaries_by_age_gender.columns = [f"({interval.left}-{interval.right}]" for interval in salaries_by_age_gender.columns]
salaries_by_age_gender

In [ ]:
# fig = px.imshow(salaries_by_age_gender, text_auto=True, title="Зависимость медианной желаемой заработной платы от возраста и пола")
# fig.update_layout(xaxis_title="Возраст", yaxis_title="Пол")
# fig.show()
# fig.write_html("plotly_graphs/12_salary_vs_age_and_gender.html")
IFrame(src='plotly_graphs/12_salary_vs_age_and_gender.html', width=1000, height=600)

По тепловой карте зависимости медианной зарплаты в рублях от пола и возраста можно заметить:
- Для мужчин пики ожиданий приходятся на интервалы 35-45 лет и 65-70 лет.
- Для женшин пики ожиданий приходятся на интервалы 40-45 лет и 70-75 лет.
- У мужчин во всех возрастных интервалах ожидания выше, кроме интервала 20-25 лет.
- Мужчины продолжают искать работу даже после 75 лет.
- До 45 лет зарплатные ожидания растут, а затем начинают снижаться.

### Зависимость зарплаты от опыта

In [ ]:
# age_exp_mask = df_with_rates["Опыт работы (месяц)"] <= ((df_with_rates["Возраст"]-14) * 12)
# salary_mask = df_with_rates['ЗП (руб)'] < 1e6

# fig = px.scatter(df_with_rates[age_exp_mask & salary_mask],
#                  x="Опыт работы (месяц)",
#                  y="ЗП (руб)",
#                  color="Пол",
#                  title="Зависимость ЗП от опыта работы")
# fig.update_traces(opacity=.4)
# fig.add_hline(y=500000, line_color="red", line_width=0.5, line_dash="dash", opacity=0.5)
# fig.add_hline(y=400000, line_color="red", line_width=0.5, line_dash="dash", opacity=0.5)
# fig.add_hline(y=350000, line_color="red", line_width=0.5, line_dash="dash", opacity=0.5)
# fig.add_hline(y=300000, line_color="red", line_width=0.5, line_dash="dash", opacity=0.5)
# fig.add_hline(y=250000, line_color="red", line_width=0.5, line_dash="dash", opacity=0.5)
# fig.add_hline(y=200000, line_color="red", line_width=0.5, line_dash="dash", opacity=0.5)
# fig.update_layout(xaxis_title="Опыт работы (месяц)", yaxis_title="ЗП (руб)")
# fig.show()
# fig.write_html("plotly_graphs/13_salary_vs_exp.html")
IFrame(src='plotly_graphs/13_salary_vs_exp.html', width=1000, height=600)

На графике зависимости ЗП от опыта работы можно заметить:
- Зарплатные ожидания растут с опытом, но после определённого момента начинают снижаться (более точно можно было это заметить на предыдущем графике).
- Наблюдаются уровни зарплатных ожиданий с шагом 50000 руб.
- У женщин, в целом, зарплатные ожидания ниже.
- После 200 месяцев количество женских резюме снижается быстрее количества мужских.

# Очистка данных

## Удалим полные дубликаты
Начнем с дубликатов в наших данных. Найдите **полные дубликаты** в таблице с резюме и удалите их.

In [ ]:
df_with_rates.shape[0] - df_with_rates.drop_duplicates(subset=df_with_rates.columns).shape[0]

In [ ]:
df_dedup = df_with_rates.drop_duplicates(subset=df_with_rates.columns).reset_index(drop=True)
df_dedup.shape

## Удалим пропуски
Займемся пропусками. Выведите информацию **о числе пропусков** в столбцах.

In [ ]:
df_dedup.info()

Итак, у нас есть пропуски в 3ех столбцах: **"Опыт работы (месяц)"**, **"Последнее/нынешнее место работы"**, **"Последняя/нынешняя должность"**. Поступим следующим образом: удалите строки, где есть пропуск в столбцах с местом работы и должностью. Пропуски в столбце с опытом работы заполните **медианным** значением.

In [ ]:
df_dedup = (df_dedup
 .dropna(subset=['Последнее/нынешнее место работы'])
 .dropna(subset=['Последняя/нынешняя должность']))

df_dedup['Опыт работы (месяц)'] = df_dedup['Опыт работы (месяц)'].fillna(df_dedup['Опыт работы (месяц)'].median())
df_dedup.shape

In [ ]:
df_dedup['Опыт работы (месяц)'].mean().round(0)

In [ ]:
df_dedup.info()

## Удалим выбросы
Мы добрались до ликвидации выбросов. Сначала очистим данные вручную. Удалите резюме, в которых указана заработная плата либо выше 1 млн. рублей, либо ниже 1 тыс. рублей.

In [ ]:
(df_dedup['ЗП (руб)'] > 1e6).sum() + (df_dedup['ЗП (руб)'] < 1e3).sum()

In [ ]:
sal_leq_1e6 = df_dedup['ЗП (руб)'] <= 1e6
sal_geq_1e3 = df_dedup['ЗП (руб)'] >= 1e3
df_dedup = df_dedup[sal_leq_1e6 & sal_geq_1e3].reset_index(drop=True)
df_dedup.shape

В процессе разведывательного анализа мы обнаружили резюме, в которых **опыт работы в годах превышал возраст соискателя**. Найдите такие резюме и удалите их из данных


In [ ]:
(df_dedup['Опыт работы (месяц)']/12 > df_dedup['Возраст']).sum()

In [ ]:
exp_greater_than_age = df_dedup['Опыт работы (месяц)']/12 > df_dedup['Возраст']
df_dedup = df_dedup[~exp_greater_than_age].reset_index(drop=True)
df_dedup.shape

## Распределение признака в логарифмическом масштабе
В результате анализа мы обнаружили потенциальные выбросы в признаке **"Возраст"**. Это оказались резюме людей чересчур преклонного возраста для поиска работы. Попробуйте построить распределение признака в **логарифмическом масштабе**. Добавьте к графику линии, отображающие **среднее и границы интервала метода трех сигм**. Напомним, сделать это можно с помощью метода axvline. Например, для построение линии среднего будет иметь вид:

`histplot.axvline(log_age.mean(), color='k', lw=2)`

В какую сторону асимметрично логарифмическое распределение? Напишите об этом в комментарии к графику.
Найдите выбросы с помощью метода z-отклонения и удалите их из данных, используйте логарифмический масштаб. Давайте сделаем послабление на **1 сигму** (возьмите 4 сигмы) в **правую сторону**.

Выведите таблицу с полученными выбросами и оцените, с каким возрастом соискатели попадают под категорию выбросов?

In [ ]:
log_age = np.log(df_dedup["Возраст"] + 1)
log_age_mean = log_age.mean()
log_age_3_sigma = log_age.std() * 3

In [ ]:
# fig = px.histogram(x=log_age,
#                    title="Распределение возраста кандидатов (в логарифмическом масштабе)")
# fig.add_vline(log_age_mean, line_width=2)
# fig.add_vline(log_age_mean + log_age_3_sigma, line_width=2, line_dash="dash")
# fig.add_vline(log_age_mean - log_age_3_sigma, line_width=2, line_dash="dash")
# fig.update_layout(xaxis_title="Возраст в логарифмическом масштабе", yaxis_title="Количество кандидатов")
# fig.show()
# fig.write_html("plotly_graphs/14_age_log.html")
IFrame(src='plotly_graphs/14_age_log.html', width=1000, height=600)

**Вопрос**: В какую сторону асимметрично логарифмическое распределение?

**Ответ**: Логарифмическое распределение асимметрично в правую сторону.

Найдите выбросы с помощью метода z-отклонения и удалите их из данных, используйте логарифмический масштаб. Давайте сделаем послабление на **1 сигму** (возьмите 4 сигмы) в **правую сторону**.



In [ ]:
def outliers_z_score(data, feature, sigmas_lower=3, sigmas_upper=3, log_scale=False):
    if log_scale:
        x = np.log(data[feature]+1)
    else:
        x = data[feature]
    mu = x.mean()
    sigma = x.std()
    lower_bound = mu - sigmas_lower * sigma
    upper_bound = mu + sigmas_upper * sigma
    outliers = data[(x < lower_bound) | (x > upper_bound)]
    cleaned = data[(x >= lower_bound) & (x <= upper_bound)]
    return outliers, cleaned

In [ ]:
outliers, cleaned = outliers_z_score(data=df_dedup, feature="Возраст", sigmas_upper=4, log_scale=True)

In [ ]:
outliers["Возраст"].value_counts()

**Вопрос**: Выведите таблицу с полученными выбросами и оцените, с каким возрастом соискатели попадают под категорию выбросов?

**Ответ**: Соискатели с возрастом 15 и 100 лет попадают под категорию статистических выбросов.